cluster.go

```diff
@@ -0,0 +1,464 @@
+package redis
+
+import (
+	"container/list"
+	"errors"
+	"io"
+	"math/rand"
+	"net"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type ClusterClient struct {
+	commandable
+
+	addrs []string
+	slots [][]string
+	conns *lruPool
+	opt   *ClusterOptions
+
+	cachemx sync.RWMutex
+	_reload uint32
+}
+
+func NewClusterClient(opt *ClusterOptions) (*ClusterClient, error) {
+	addrs, err := opt.getAddrs()
+	if err != nil {
+		return nil, err
+	}
+
+	client := &ClusterClient{
+		addrs:   addrs,
+		conns:   newLRUPool(opt),
+		opt:     opt,
+		_reload: 1,
+	}
+	client.commandable.process = client.process
+	client.reloadIfDue()
+	return client, nil
+}
+
+// Closes the pool
+func (c *ClusterClient) Close() error {
+	c.cachemx.Lock()
+	defer c.cachemx.Unlock()
+
+	return c.reset()
+}
+
+// GetMasterAddrBySlot finds the current master address for a given hash slot
+func (c *ClusterClient) GetMasterAddrBySlot(hashSlot int) string {
```

Comment:
Can we unexport this?

API seems to be low-level, e.g. it is unclear how user should obtain hashSlot.